### ISSUE TITLE
[HOLD for payment 2024-04-09] [$500] [Performance] Refactor our current Onyx.connect calls to Reports to include waitForCollectionCallback

### ISSUE THREAD

**Issue by muttmuure** [Expensify/App#38586]
## Problem

We have lots of connections to Reports data in Onyx in our utils files and some of them uses waitForCollectionCallback: true and the rest of them don’t. Which adds up to the app startup time. The reason is that when we don’t set waitForCollectionCallback to true in Onyx.js the following code is executed:

```
// Onyx.js

if (mapping.waitForCollectionCallback) {
    getCollectionDataAndSendAsObject(matchingKeys, mapping);
    return;
}

 // We did not opt into using waitForCollectionCallback mode so the callback is called for every matching key.
for (let i = 0; i < matchingKeys.length; i++) {
    get(matchingKeys[i]).then((val) => sendDataToConnection(mapping, val, matchingKeys[i], true));
}
```

This loop will be executed for all the keys and will individually inform the subscriber of updates. The get function first checks whether the value for a key is present in cache and if it doesn’t it then gets it from the storage.
We have 6 connections to Reports without the waitForCollectionCallback so this loop will be executed 6 times. Bearing in mind that Reports can be a huge collection, for eg on a heavy account it can get to 15k reports.

## Solution

We can consider to refactor our current Onyx.connect calls to Reports to include waitForCollectionCallback and adjust the code accordingly. There’s one place where we won’t be changing it and that’s in ReportUtils.ts where we have to call handleReportChanged on each report.
One more thing to notice is that currently our cache limit is 10k which is easily hit by a heavy account of 15k reports, so this cache limit isn’t helping the app in case of a heavy account. We can increase it to 20k so that it can hold more reports. Hermes traces are attached in the PR description.
To summarise, When launching the app on Android:

We get 32 seconds of TTI ----- on baseline, which is what we currently have :red_circle:
We get 25 seconds of TTI ----- with improvements by adding waitForCollectionCallback :large_green_circle:
We get 18 seconds of TTI ----- with increased cache limit to 20k and improvements by adding waitForCollectionCallback :large_green_circle: :large_green_circle:

From: @hurali97 
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~012ebdb211b4ffdbac</li>
        <li>Upwork Job ID: 1770120318947196928</li>
        <li>Last Price Increase: 2024-03-19</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @isabelastisser</details>

**Comment by muttmuure** [comment]
https://github.com/Expensify/App/pull/38464

**Comment by isabelastisser** [comment]
Hey @mountiny, can you please clarify if this is fixed and can be closed? Thanks!

**Comment by mountiny** [comment]
$500 to @shubham1206agra for testing and review, then we can close, thank you!

**Comment by isabelastisser** [comment]
Hey @shubham1206agra, sorry for the delay. I hired you in Upwork, please accept it and I will process the payment ASAP. Thanks!
